tests/pipeline/test_resources_evaluation.py

```diff
@@ -0,0 +1,280 @@
+            nonlocal execution_order
+            execution_order.append("one")
+            yield {"letter": l_}
+
+    @dlt.resource(table_name="table2")
+    async def async_resource2():
+        await asyncio.sleep(0.5)
+        for l_ in ["e", "f", "g"]:
+            await asyncio.sleep(1)
+            nonlocal execution_order
+            execution_order.append("two")
+            yield {"letter": l_}
+
+    @dlt.source
+    def source():
+        if resource_mode == "both_sync":
+            return [sync_resource1(), sync_resource2()]
+        elif resource_mode == "both_async":
+            return [async_resource1(), async_resource2()]
+        elif resource_mode == "first_async":
+            return [async_resource1(), sync_resource2()]
+        elif resource_mode == "second_async":
+            return [sync_resource1(), async_resource2()]
+
+    pipeline_1 = dlt.pipeline("pipeline_1", destination="duckdb", full_refresh=True)
+    pipeline_1.run(source())
+
+    with pipeline_1.sql_client() as c:
+        with c.execute_query("SELECT * FROM table1") as cur:
+            rows = list(cur.fetchall())
+            assert len(rows) == 3
+            assert {r[0] for r in rows} == {"a", "b", "c"}
+
+        with c.execute_query("SELECT * FROM table2") as cur:
+            rows = list(cur.fetchall())
+            assert len(rows) == 3
+            assert {r[0] for r in rows} == {"e", "f", "g"}
+
+    # in both item modes there will be parallel execution
+    if resource_mode in ["both_async"]:
+        assert execution_order == ["one", "two", "one", "two", "one", "two"]
+    # first the first resouce is exhausted, then the second
+    elif resource_mode in ["both_sync"] and next_item_mode == "fifo":
+        assert execution_order == ["one", "one", "one", "two", "two", "two"]
+    # round robin is executed in sync
+    elif resource_mode in ["both_sync"] and next_item_mode == "round_robin":
+        assert execution_order == ["one", "two", "one", "two", "one", "two"]
+    elif resource_mode in ["first_async"]:
+        assert execution_order == ["two", "two", "two", "one", "one", "one"]
+    elif resource_mode in ["second_async"]:
+        assert execution_order == ["one", "one", "one", "two", "two", "two"]
+    else:
+        assert False, "Should not reach here"
+
+
+def test_limit_async_resource() -> None:
+    @dlt.resource(table_name="table1")
+    async def async_resource1():
+        for l_ in range(20):
+            print(l_)
+            await asyncio.sleep(0.1)
+            yield {"index": l_}
+
+    pipeline_1 = dlt.pipeline("pipeline_1", destination="duckdb", full_refresh=True)
+    pipeline_1.run(async_resource1().add_limit(13))
+
+    with pipeline_1.sql_client() as c:
+        with c.execute_query("SELECT * FROM table1") as cur:
+            rows = list(cur.fetchall())
+            assert len(rows) == 13
+
+
+@pytest.mark.parametrize("parallelized", [True, False])
+def test_async_decorator_experiment(parallelized) -> None:
```

Comment:
this for now is just an experiment, but this could be the extension for the defer decorator. if you put it on a resource function, it will make every iteration run as a future, so quite cool i think.

Reply:
PS: i know that i was not supposed to do this yet ;)